### PR TITLE
Add XOX70 firmware

### PR DESCRIPTION
**Pull request by ToastyStoemp**
## Description

Add base firmware for XOX70 Hotswap and Solder

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/wolf/xox70/hotswap/config.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+#pragma once
+
+#include "config_common.h"
+
```
Comment (drashna) [inline keyboards/wolf/xox70/hotswap/config.h:20]:
```suggestion
```

File: keyboards/wolf/xox70/hotswap/readme.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# XOX70_Hotswap
+
+![X0X70](https://cdn.shopify.com/s/files/1/0505/0328/6952/products/WY.jpg?v=1661602119&width=640)
```
Comment (drashna) [inline keyboards/wolf/xox70/hotswap/readme.md:3]: image should be rehosted on an image host. Not shopify.

File: keyboards/wolf/xox70/hotswap/keymaps/via/keymap.c [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            xox.caps.v = value_data[0];
+            break;
+        }
+        case id_caps_indicator_color: {
+            xox.caps.h = value_data[0];
+            xox.caps.s = value_data[1];
+            break;
+        }
+        case id_leds_state: {
+            xox.leds = value_data[0];
```
Comment (drashna) [inline keyboards/wolf/xox70/hotswap/keymaps/via/keymap.c:151]: This could and probably should be translated to use the rgb_matrix_set_flags function.

Comment (ToastyStoemp) [reply keyboards/wolf/xox70/hotswap/keymaps/via/keymap.c]: I've removed the configuration from the VIA RGB to be more compliant with current QMK's stance on too much code in the keymaps that's duplicate to many other boards.
I appreciate you looking at this

**Review by drashna — APPROVED** [review]

File: keyboards/wolf/xox70/hotswap/config.h [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+#pragma once
+
+/* Define RGB */
+#ifdef RGB_MATRIX_ENABLE
+
```
Comment (fauxpark) [inline keyboards/wolf/xox70/hotswap/config.h:21]:
```suggestion
```

File: keyboards/wolf/xox70/hotswap/config.h [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+#define ENABLE_RGB_MATRIX_SPLASH
+#define ENABLE_RGB_MATRIX_MULTISPLASH
+#define ENABLE_RGB_MATRIX_SOLID_SPLASH
+#define ENABLE_RGB_MATRIX_SOLID_MULTISPLASH
+
+#define EECONFIG_KB_DATA_SIZE 5
+
+#define CAPS_INDICATOR_INDEX 37
+
+#endif
```
Comment (fauxpark) [inline keyboards/wolf/xox70/hotswap/config.h:91]:
```suggestion
```

File: keyboards/wolf/xox70/hotswap/config.h [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+#define RGB_MATRIX_LED_COUNT 72
+#define WS2812_DI_PIN GP12
+
+#define RGB_DISABLE_WHEN_USB_SUSPENDED
+#define RGB_MATRIX_KEYPRESSES 
+#define RGB_MATRIX_FRAMEBUFFER_EFFECTS
+#define RGB_MATRIX_MAXIMUM_BRIGHTNESS 120
+
+#define RGB_MATRIX_STARTUP_MODE RGB_MATRIX_CYCLE_ALL
```
Comment (fauxpark) [inline keyboards/wolf/xox70/hotswap/config.h:32]:
```suggestion
#define RGB_MATRIX_DEFAULT_MODE RGB_MATRIX_CYCLE_ALL
```

File: keyboards/wolf/xox70/hotswap/readme.md [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+The following is the QMK Firmware for the XOX70_Hotswap PCB.
+
+* Keyboard Maintainer: [Wolf](https://github.com/ToastyStoemp)
+* Hardware Supported: XOX70 Hotswap PCB
+* Hardware Availability: https://thocc.supply/products/xox70
+
+Make example for this keyboard (after setting up your build environment):
+
+	make wolf/xox70/hotswap:default
```
Comment (fauxpark) [inline keyboards/wolf/xox70/hotswap/readme.md:13]:
```suggestion
    make wolf/xox70/hotswap:default
```

File: keyboards/wolf/xox70/hotswap/rules.mk [hunk 8]
```diff
@@ -0,0 +1 @@
+WS2812_DRIVER = vendor
```
Comment (fauxpark) [inline keyboards/wolf/xox70/hotswap/rules.mk:1]: This goes in info.json as `ws2812.driver`.

File: keyboards/wolf/xox70/solder/config.h [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+#define BACKLIGHT_PWM_DRIVER PWMD0
+#define BACKLIGHT_PWM_CHANNEL RP2040_PWM_CHANNEL_B
+
+/* Define RGB */
+#ifdef RGBLIGHT_ENABLE
+
+#define WS2812_PIO_USE_PIO1
+
+#define RGBLIGHT_DEFAULT_MODE RGBLIGHT_MODE_RAINBOW_SWIRL
+#endif
```
Comment (fauxpark) [inline keyboards/wolf/xox70/solder/config.h:28]:
```suggestion
#define WS2812_PIO_USE_PIO1

#define RGBLIGHT_DEFAULT_MODE RGBLIGHT_MODE_RAINBOW_SWIRL
```

File: keyboards/wolf/xox70/solder/halconf.h [hunk 10]
```diff
@@ -0,0 +1,8 @@
+// Copyright 2022 MatteoDM <matteodalmo@gmail.com>
+// SPDX-License-Identifier: GPL-2.0-or-later
+#pragma once
+
+#define HAL_USE_PWM TRUE
+
+
```
Comment (fauxpark) [inline keyboards/wolf/xox70/solder/halconf.h:7]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/wolf/xox70/solder/info.json [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+            "twinkle": true
+        },
+        "led_count": 2,
+        "max_brightness": 120,
+        "pin": "GP12",
+        "sleep": true
+    },
+    "ws2812": {
+        "pin": "GP12",
+        "driver" : "vendor"
```
Comment (fauxpark) [inline keyboards/wolf/xox70/solder/info.json:45]:
```suggestion
        "driver": "vendor"
```

File: keyboards/wolf/xox70/solder/info.json [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+            "rainbow_mood": true,
+            "rainbow_swirl": true,
+            "rgb_test": true,
+            "snake": true,
+            "static_gradient": true,
+            "twinkle": true
+        },
+        "led_count": 2,
+        "max_brightness": 120,
+        "pin": "GP12",
```
Comment (fauxpark) [inline keyboards/wolf/xox70/solder/info.json:40]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/wolf/xox70/hotswap/config.h
Comment (dunk2k) [inline keyboards/wolf/xox70/hotswap/config.h:1]: Can settings in this file that have data-driven equivalent be moved to `info.json` please?

File: keyboards/wolf/xox70/hotswap/readme.md [hunk 13]
```diff
@@ -0,0 +1,18 @@
+# XOX70_Hotswap
+
+![X0X70](https://i.imgur.com/OSW2c6L.png)
```
Comment (dunk2k) [inline keyboards/wolf/xox70/hotswap/readme.md:3]:
```suggestion
![X0X70](https://i.imgur.com/OSW2c6Lh.png)
```

File: keyboards/wolf/xox70/hotswap/readme.md [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+* Hardware Availability: https://thocc.supply/products/xox70
+
+Make example for this keyboard (after setting up your build environment):
+
+    make wolf/xox70/hotswap:default
+
+See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
+
+## Bootloader Enter the bootloader in 1 ways: 
+* **Bootmagic reset**: Hold down the key ESC key and plug in the keyboard (Top Left most switch)
```
Comment (dunk2k) [inline keyboards/wolf/xox70/hotswap/readme.md:18]:
```suggestion
## Bootloader Enter the bootloader in 2 ways: 
* **Bootmagic reset**: Hold down the key ESC key and plug in the keyboard (Top Left most switch)
* **Keycode in layout**: Press the key mapped to `QK_BOOT` if it is available
```

File: keyboards/wolf/xox70/solder/readme.md [hunk 15]
```diff
@@ -0,0 +1,20 @@
+# XOX70_Solder
+
+![X0X70](https://i.imgur.com/OSW2c6L.png)
```
Comment (dunk2k) [inline keyboards/wolf/xox70/solder/readme.md:3]:
```suggestion
![X0X70](https://i.imgur.com/OSW2c6Lh.png)
```

File: keyboards/wolf/xox70/solder/readme.md [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+* Keyboard Maintainer: [ToastyStoemp](https://github.com/ToastyStoemp)
+
+Make example for this keyboard (after setting up your build environment):
+
+    make wolf/xox70/solder:default
+
+See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
+
+## Bootloader Enter the bootloader in 1 ways: 
+* **Bootmagic reset**: Hold down the key ESC key and plug in the keyboard (Top Left most switch)
```
Comment (dunk2k) [inline keyboards/wolf/xox70/solder/readme.md:20]:
```suggestion
## Bootloader Enter the bootloader in 2 ways: 
* **Bootmagic reset**: Hold down the key ESC key and plug in the keyboard (Top Left most switch)
* **Keycode in layout**: Press the key mapped to `QK_BOOT` if it is available
```

File: keyboards/wolf/xox70/solder/info.json [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+        "driver": "vendor"
+    },
+    "url": "",
+    "usb": {
+        "device_version": "1.0.0",
+        "pid": "0x0083",
+        "vid": "0x5453"
+    },
+    "layouts": {
```
Comment (dunk2k) [inline keyboards/wolf/xox70/solder/info.json:52]: experience leads to me confirm with contributor - does this board have more than one key arrangement?

**Review by dunk2k — CHANGES_REQUESTED** [review]

File: keyboards/wolf/xox70/hotswap/config.h [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+#undef ENABLE_RGB_MATRIX_PIXEL_FRACTAL
+
+// Framebuffer effects; can be enabled only if RGB_MATRIX_FRAMEBUFFER_EFFECTS
+// is defined.  Both of these effects currently don't work properly when the
+// key matrix does not match the physical layout, so they are disabled.
+#undef ENABLE_RGB_MATRIX_TYPING_HEATMAP
+#undef ENABLE_RGB_MATRIX_DIGITAL_RAIN
+
+// Reactive effects; can be enabled only if at least one of
+// RGB_MATRIX_KEYPRESSES or RGB_MATRIX_KEYRELEASES is defined.
```
Comment (drashna) [inline keyboards/wolf/xox70/hotswap/config.h:69]:
```suggestion
```

File: keyboards/wolf/xox70/hotswap/info.json [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+        "cols": ["GP24", "GP23", "GP22", "GP21", "GP20", "GP19", "GP18", "GP17", "GP16", "GP29", "GP28", "GP27", "GP26", "GP25", "GP2", "GP1", "GP0"],
+        "rows": ["GP4", "GP5", "GP6", "GP7", "GP3"]
+    },
+    "processor": "RP2040",
+    "ws2812": {
+        "pin": "GP12",
+        "driver" : "vendor"
+    },
+    "rgb_matrix": {
+        "driver": "WS2812",
```
Comment (drashna) [inline keyboards/wolf/xox70/hotswap/info.json:27]:
```suggestion
        "driver": "ws2812",
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/wolf/xox70/hotswap/info.json [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+        ]
+    },
+    "url": "",
+    "usb": {
+        "device_version": "1.0.0",
+        "pid": "0x0082",
+        "vid": "0x5453"
+    },
+    "layouts": {
+        "LAYOUT": {
```
Comment (noroadsleft) [inline keyboards/wolf/xox70/hotswap/info.json:110]:
```suggestion
        "LAYOUT_tkl_nofrow_ansi_tsangan": {
```

File: keyboards/wolf/xox70/hotswap/keymaps/default/keymap.c [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+    [0] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/wolf/xox70/hotswap/keymaps/default/keymap.c:20]:
```suggestion
    [0] = LAYOUT_tkl_nofrow_ansi_tsangan(
```

File: keyboards/wolf/xox70/hotswap/keymaps/default/keymap.c [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+        QK_GESC, KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC,       KC_INS,  KC_HOME, KC_PGUP,
+        KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_BSLS,       KC_DEL,  KC_END,  KC_PGDN,
+        KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,          KC_ENT,
+        KC_LSFT,          KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,          KC_RSFT,                KC_UP,
+        KC_LCTL, KC_LGUI, KC_LALT,                            KC_SPC,                                      KC_RALT, MO(1),   KC_RCTL,       KC_LEFT, KC_DOWN, KC_RGHT
+    ),
+
+    [1] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/wolf/xox70/hotswap/keymaps/default/keymap.c:28]:
```suggestion
    [1] = LAYOUT_tkl_nofrow_ansi_tsangan(
```

File: keyboards/wolf/xox70/hotswap/keymaps/via/keymap.c [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+    [0] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/wolf/xox70/hotswap/keymaps/via/keymap.c:21]:
```suggestion
    [0] = LAYOUT_tkl_nofrow_ansi_tsangan(
```

File: keyboards/wolf/xox70/hotswap/keymaps/via/keymap.c [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+        QK_GESC, KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC,   KC_INS,  KC_HOME, KC_PGUP,
+        KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_BSLS,   KC_DEL,  KC_END,  KC_PGDN,
+        KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,          KC_ENT,
+        KC_LSFT,          KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,          KC_RSFT,            KC_UP,
+        KC_LCTL, KC_LGUI, KC_LALT,                            KC_SPC,                                      KC_RALT, MO(1),   KC_RCTL,   KC_LEFT, KC_DOWN, KC_RGHT
+    ),
+
+    [1] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/wolf/xox70/hotswap/keymaps/via/keymap.c:29]:
```suggestion
    [1] = LAYOUT_tkl_nofrow_ansi_tsangan(
```

File: keyboards/wolf/xox70/solder/info.json [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+        "LAYOUT": {
```
Comment (noroadsleft) [inline keyboards/wolf/xox70/solder/info.json:53]:
```suggestion
        "LAYOUT_all": {
```

File: keyboards/wolf/xox70/solder/info.json [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+                { "matrix": [2, 4], "x": 4.75, "y": 2 },
+                { "matrix": [2, 5], "x": 5.75, "y": 2 },
+                { "matrix": [2, 6], "x": 6.75, "y": 2 },
+                { "matrix": [2, 7], "x": 7.75, "y": 2 },
+                { "matrix": [2, 8], "x": 8.75, "y": 2 },
+                { "matrix": [2, 9], "x": 9.75, "y": 2 },
+                { "matrix": [2, 10], "x": 10.75, "y": 2 },
+                { "matrix": [2, 11], "x": 11.75, "y": 2 },
+                { "matrix": [2, 12], "x": 19.5, "y": 2 },
+                { "matrix": [2, 13], "w": 2.25, "x": 12.75, "y": 2 },
```
Comment (noroadsleft) [inline keyboards/wolf/xox70/solder/info.json:103]:
```suggestion
                { "matrix": [2, 12], "x": 12.75, "y": 2 },
                { "matrix": [2, 13], "w": 1.25, "x": 13.75, "y": 2 },
```

File: keyboards/wolf/xox70/solder/info.json [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+                { "matrix": [4, 0], "w": 1.5, "x": 0, "y": 4 },
+                { "matrix": [4, 1], "x": 1.5, "y": 4 },
+                { "matrix": [4, 2], "w": 1.5, "x": 2.5, "y": 4 },
+                { "matrix": [4, 3], "w": 3, "x": 4, "y": 4 },
+                { "matrix": [4, 4], "w": 7, "x": 4, "y": 4 },
+                { "matrix": [4, 9], "w": 3, "x": 8, "y": 4 },
+                { "matrix": [4, 10], "w": 1.25, "x": 10, "y": 5.25 },
+                { "matrix": [4, 11], "w": 1.5, "x": 11, "y": 4 },
+                { "matrix": [4, 12], "x": 12.5, "y": 4 },
+                { "matrix": [4, 13], "w": 1.5, "x": 13.5, "y": 4 },
```
Comment (noroadsleft) [inline keyboards/wolf/xox70/solder/info.json:128]: Cleans up the Configurator rendering on the bottom row.

```suggestion
                { "matrix": [4, 0], "w": 1.25, "x": 0, "y": 4 },
                { "matrix": [4, 1], "w": 1.25, "x": 1.25, "y": 4 },
                { "matrix": [4, 2], "w": 1.25, "x": 2.5, "y": 4 },
                { "matrix": [4, 3], "w": 2, "x": 3.75, "y": 4 },
                { "matrix": [4, 4], "w": 2.25, "x": 5.75, "y": 4 },
                { "matrix": [4, 9], "w": 2, "x": 8, "y": 4 },
                { "matrix": [4, 10], "w": 1.25, "x": 10, "y": 5.25 },
                { "matrix": [4, 11], "w": 1.25, "x": 11.25, "y": 4 },
                { "matrix": [4, 12], "w": 1.25, "x": 12.5, "y": 4 },
                { "matrix": [4, 13], "w": 1.25, "x": 13.75, "y": 4 },
```

File: keyboards/wolf/xox70/solder/keymaps/default/keymap.c [hunk 28]
```diff
@@ -0,0 +1,27 @@
...
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+    [0] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/wolf/xox70/solder/keymaps/default/keymap.c:20]:
```suggestion
    [0] = LAYOUT_all(
```

File: keyboards/wolf/xox70/solder/keymaps/via/keymap.c [hunk 29]
```diff
@@ -0,0 +1,27 @@
...
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+    [0] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/wolf/xox70/solder/keymaps/via/keymap.c:20]:
```suggestion
    [0] = LAYOUT_all(
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]
GitHub tip: You can apply multiple suggestions to a single commit by using the [Files Changed tab](https://github.com/qmk/qmk_firmware/pull/21023/files).

File: keyboards/wolf/xox70/solder/info.json (hunk 17 above)
Comment (noroadsleft) [reply keyboards/wolf/xox70/solder/info.json:52]: It does:

https://cdn.shopify.com/s/files/1/0505/0328/6952/files/XOX_Layout.5.jpg:
image

**Pull request closed without merging**